moveit_ros/planning/planning_request_adapter_plugins/src/fix_start_state_collision.cpp

```diff
@@ -115,37 +127,106 @@ class FixStartStateCollision : public planning_request_adapter::PlanningRequestA
+        else if (!req.goal_constraints[0].orientation_constraints.empty())
+          ee_link_model = start_state.getLinkModel(req.goal_constraints[0].orientation_constraints[0].link_name);
+      }
+      else
+      {
+        ROS_ERROR("No goal constraint set; aborting FixStartStateCollision.");
+        return planner(planning_scene, req, res);
+      }
+
+      // Collect the links that will be checked
+      std::vector<const moveit::core::LinkModel*> link_models;
+      link_models.resize(jmodels.size());
+      for (std::size_t i = jmodels.size() - 1; i > 0; --i)
+        link_models[i] = jmodels[i]->getChildLinkModel();
+
+      if (std::find(link_models.begin(), link_models.end(), ee_link_model) == link_models.end())
+      {  // If the end effector is a separate link, add it to the front of the vector
+        link_models.resize(jmodels.size() + 1);
+        link_models[0] = ee_link_model;
+        for (std::size_t i = jmodels.size() - 1; i > 0; --i)
+          link_models[i + 1] = jmodels[i]->getChildLinkModel();
+      }
+
+      // Store the unchanged link positions
+      EigenSTL::vector_Isometry3d original_link_poses;
+      original_link_poses.resize(link_models.size());
+      Eigen::Isometry3d new_link_pose;
+
+      for (std::size_t i = 0; i < link_models.size(); ++i)
+        original_link_poses[i] = start_state.getGlobalLinkTransform(link_models[i]);
+
+      double dist, max_dist;
+      double local_jiggle_fraction = jiggle_fraction_;
       bool found = false;
-      for (int c = 0; !found && c < sampling_attempts_; ++c)
+      for (int c = 0; !found && c < max_sampling_attempts_; ++c)
       {
+        // Change each joint slightly ("jiggle")
         for (std::size_t i = 0; !found && i < jmodels.size(); ++i)
         {
           std::vector<double> sampled_variable_values(jmodels[i]->getVariableCount());
-          const double* original_values = prefix_state->getJointPositions(jmodels[i]);
+          const double* original_values = original_state->getJointPositions(jmodels[i]);
           jmodels[i]->getVariableRandomPositionsNearBy(rng, &sampled_variable_values[0], original_values,
                                                        jmodels[i]->getMaximumExtent() * jiggle_fraction_);
           start_state.setJointPositions(jmodels[i], sampled_variable_values);
-          collision_detection::CollisionResult cres;
-          planning_scene->checkCollision(creq, cres, start_state);
-          if (!cres.collision)
+        }
+
+        // Confirm that no links move far after jiggling
+        max_dist = -1;
+        bool euclidean_distance_ok = true;
+        for (std::size_t i = 0; !found && i < link_models.size(); ++i)
+        {
+          new_link_pose = start_state.getGlobalLinkTransform(link_models[i]);
+          dist = (new_link_pose.translation() - original_link_poses[i].translation()).squaredNorm();
+
+          if (dist > max_dist)
+            max_dist = dist;
```

Comment:
`max_dist` isn't used yet. But I think, it might be useful: My suggestion is to traverse all links and compute the largest distance. If any link moves too far, i.e. if `max_dist` is too large, try to downscale the jiggling motion. For small joint motions, the relationship between joint and Cartesian deltas is linear. Thus downscaling the joint delta by the factor `0.9 * sqrt(max_dist) / max_jiggle_distance` should result in a valid Cartesian motion. The factor 0.9 is a safety margin to handle nonlinearities. Maybe you can even drop it.